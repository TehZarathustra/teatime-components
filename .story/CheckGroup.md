## Preview


### size s

```
<div>
  some text for the baseline
  <CheckGroup
    cols={3}
    name='checkGroup-s'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='s'/>
  {' '}
  <CheckGroup
    cols={3}
    disabled={true}
    name='checkGroup-s-d'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='s'/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <CheckGroup
    cols={3}
    defaultValue={[
      'kawasaki',
      'vespa'
    ]}
    name='checkGroup-m'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='m'/>
  {' '}
  <CheckGroup
    cols={3}
    disabled={true}
    name='checkGroup-m-d'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='m'/>
</div>
```
