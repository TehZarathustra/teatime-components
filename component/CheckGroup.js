'use strict';

const { PropTypes } = require('react');
const CheckGroup = require('../view/CheckGroup');
const simplify = require('../mixin/simplify');

module.exports = simplify(
  CheckGroup,
  (styles, { size }) => ({
    styles: styles[size],
  }),
  {
    m: require('../style/check-group/check-group-m.css'),
    l: require('../style/check-group/check-group-l.css'),
  },
  {
    size: 'm',
  },
  {
    size: PropTypes.oneOf([
      'm',
      'l',
    ]),
  }
);
