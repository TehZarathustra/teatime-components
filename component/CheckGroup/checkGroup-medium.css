.checkGroup
{
  composes: checkGroup from './checkGroup.css';
}

.column
{
  composes: column from './checkGroup.css';
}

.check
{
  composes: check from './checkGroup.css';

  margin: 0 0 15px;
}
