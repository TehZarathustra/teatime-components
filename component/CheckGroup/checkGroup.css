.checkGroup
{
  display: flex;

  flex-flow: row nowrap;
  justify-content: space-between;
}

.column
{
  display: block;

  flex: 1 1 0;
}

.check
{
  display: block;
}

.check:last-child
{
  margin: 0;
}
