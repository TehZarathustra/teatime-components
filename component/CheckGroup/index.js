/* @flow */
import {buildValueHash, chunk, isControlled, map, noop, valuesFromHash} from '../../lib/util';
import Check from '../Check';
import React, {Component} from 'react';
import classNames from 'classnames';

import checkGroupLarge from './checkGroup-large.css';
import checkGroupMedium from './checkGroup-medium.css';

type Item = {
  label: string,
  value: string,
};

type Props = {
  className?: string,
  cols: number,
  defaultValue?: Array<string>,
  disabled: boolean,
  name: string,
  onChange: Function,
  options: Array<Item>,
  size: 'l' | 'm',
  style?: Object,
  styles?: Object,
  value?: Array<string>,
};

type CheckProps = {
  className: string,
  disabled: boolean,
  key: string,
  name: string,
  onChange: Function,
  size: 'l' | 'm',
  value: string,
};

type ColumnProps = {
  children: any,
  className: string,
  key: string,
};

const cssModules = {
  l: checkGroupLarge,
  m: checkGroupMedium,
};

class CheckGroup extends Component {
  props: Props

  static defaultProps: {
    cols: number,
    disabled: boolean,
    onChange: Function,
    size: 'l' | 'm',
  }

  state: {
    values: Object,
  }

  css: {
    checkGroup: string,
    column: string,
    check: string,
  }

  _controlled: boolean

  constructor(props: Props) {
    super(props);

    this._controlled = isControlled(props);

    const value = this._controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      values: buildValueHash(props.options, value),
    };

    this.css = props.styles || cssModules[props.size];

    (this:any)._onChange = this._onChange.bind(this);
  }

  componentWillReceiveProps(nextProps: Props) {
    this._controlled = isControlled(nextProps);

    if (
      this._controlled &&
      (
        nextProps.options !== this.props.options ||
        nextProps.value !== this.props.value
      )
    ) {
      const {options, value} = nextProps;

      this.setState({
        values: buildValueHash(options, value),
      });
    }

    this.css = nextProps.styles || cssModules[nextProps.size];
  }

  focus() {} // noop

  _computeItems(options: Array<Item>) {
    const {disabled, name, size} = this.props;
    const {values} = this.state;
    const {css} = this;

    return map(item => this.renderCheck(Object.assign({}, item, {
      checked: values[item.value],
      className: css.check,
      disabled: item.disabled || disabled,
      key: item.value,
      name,
      onChange: this._onChange,
      size,
    })), options);
  }

  _computeColumns(checks: Array<mixed>) {
    const {cols} = this.props;
    const {css} = this;

    const chunkSize = Math.ceil(checks.length / cols)
    const columns = chunk(checks, chunkSize);

    let index = 0;

    return map(children => this.renderColumn({
      children,
      className: css.column,
      key: String(++index),
    }), columns);
  }

  _onChange(e: SyntheticInputEvent, {checked, value}: Object) {
    const {name, onChange} = this.props;
    const {values} = this.state;
    const nextValues = Object.assign({}, values, {[value]: checked});

    if (!this._controlled) {
      this.setState({values: nextValues});
    }

    const nextValue = valuesFromHash(nextValues);
    onChange(e, {name, value: nextValue});
  }

  render() {
    const {className, style} = this.props;
    const {css} = this;

    const items = this._computeItems(this.props.options);
    const columns = this._computeColumns(items);

    return (
      <div className={classNames(css.checkGroup, className)} style={style}>
        {columns}
      </div>
    );
  }

  renderCheck(checkProps: CheckProps) {
    return (
      <Check {...checkProps}/>
    );
  }

  renderColumn(columnProps: ColumnProps) {
    return (
      <div {...columnProps}/>
    );
  }
}

CheckGroup.defaultProps = {
  cols: 1,
  disabled: false,
  onChange: noop,
  size: 'm',
};

export default CheckGroup;
