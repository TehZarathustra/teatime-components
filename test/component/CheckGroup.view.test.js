'use strict';

import CheckGroup from '../../component/CheckGroup';
import React from 'react';
import renderer from 'react-test-renderer';

test('defaultProp #size should have a value `m`', () => {
  expect(CheckGroup.defaultProps.size).toBe('m');
});

test('Renders correctly with default props', () => {
  const tree = renderer.create(
    <CheckGroup
      name='control'
      options={[
        {label: 'AA', value: 'aa'},
        {label: 'BB', value: 'bb'},
        {label: 'CC', value: 'cc'},
      ]}/>
  );

  expect(tree).toMatchSnapshot();
});

test('Renders correctly with defined props', () => {
  const tree = renderer.create(
    <CheckGroup
      className='outer'
      cols={2}
      disabled
      id='controls-id'
      name='control'
      options={[
        {label: 'AA', value: 'aa'},
        {label: 'BB', value: 'bb'},
        {label: 'CC', value: 'cc'},
      ]}
      value={[
        'bb',
        'cc',
      ]}/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
