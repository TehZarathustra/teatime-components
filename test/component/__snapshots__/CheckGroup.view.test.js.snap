// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with default props 1`] = `
<div
  className="checkGroup"
  style={undefined}
>
  <div
    className="column"
  >
    <span
      className="check check"
      style={undefined}
    >
      <input
        checked={false}
        className="native"
        defaultChecked={undefined}
        disabled={false}
        id="_teatime1"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="aa"
      />
      <label
        className="control"
        htmlFor="_teatime1"
      />
      <label
        className="label"
        htmlFor="_teatime1"
      >
        AA
      </label>
    </span>
    <span
      className="check check"
      style={undefined}
    >
      <input
        checked={false}
        className="native"
        defaultChecked={undefined}
        disabled={false}
        id="_teatime2"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="bb"
      />
      <label
        className="control"
        htmlFor="_teatime2"
      />
      <label
        className="label"
        htmlFor="_teatime2"
      >
        BB
      </label>
    </span>
    <span
      className="check check"
      style={undefined}
    >
      <input
        checked={false}
        className="native"
        defaultChecked={undefined}
        disabled={false}
        id="_teatime3"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="cc"
      />
      <label
        className="control"
        htmlFor="_teatime3"
      />
      <label
        className="label"
        htmlFor="_teatime3"
      >
        CC
      </label>
    </span>
  </div>
</div>
`;

exports[`Renders correctly with defined props 1`] = `
<div
  className="checkGroup outer"
  style={undefined}
>
  <div
    className="column"
  >
    <span
      className="check check"
      style={undefined}
    >
      <input
        checked={false}
        className="native"
        defaultChecked={undefined}
        disabled={true}
        id="_teatime4"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="aa"
      />
      <label
        className="control"
        htmlFor="_teatime4"
      />
      <label
        className="label"
        htmlFor="_teatime4"
      >
        AA
      </label>
    </span>
    <span
      className="check check"
      style={undefined}
    >
      <input
        checked={true}
        className="native"
        defaultChecked={undefined}
        disabled={true}
        id="_teatime5"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="bb"
      />
      <label
        className="control"
        htmlFor="_teatime5"
      />
      <label
        className="label"
        htmlFor="_teatime5"
      >
        BB
      </label>
    </span>
  </div>
  <div
    className="column"
  >
    <span
      className="check check"
      style={undefined}
    >
      <input
        checked={true}
        className="native"
        defaultChecked={undefined}
        disabled={true}
        id="_teatime6"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="cc"
      />
      <label
        className="control"
        htmlFor="_teatime6"
      />
      <label
        className="label"
        htmlFor="_teatime6"
      >
        CC
      </label>
    </span>
  </div>
</div>
`;
